src/wallet/rpcwallet.cpp

```diff
@@ -2582,6 +2584,262 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        if (options.size() > 3)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Too many optional parameters");
+        RPCTypeCheckObj(options,
+            {
+                {"confTarget", UniValueType(UniValue::VNUM)},
+                {"totalFee", UniValueType(UniValue::VNUM)},
+                {"maxFee", UniValueType(UniValue::VNUM)},
+                {"force", UniValueType(UniValue::VBOOL)},
+            },
+            true, true);
+
+        if (options.exists("confTarget")) {
+            newConfirmTarget = options["confTarget"].get_int();
+            if (newConfirmTarget <= 0) // upper-bound will be checked by estimatefee/smartfee
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid confTarget (cannot be <= 0)");
+        }
+        if (options.exists("maxFee")) {
+            CAmount proposedMaxFee = options["maxFee"].get_int();
+            if (proposedMaxFee <= 0)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid maxFee (cannot be <= 0)");
+            else if (proposedMaxFee > maxTxFee)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid maxFee (cannot be higher than the global maxTxFee)");
+            bumpfeeMaxTxFee = proposedMaxFee;
+        }
+        if (options.exists("totalFee")) {
+            totalFee = options["totalFee"].get_int();
+            if (totalFee <= 0)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid totalFee (cannot be <= 0)");
+            else if (totalFee > bumpfeeMaxTxFee)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid totalFee (cannot be higher than bumpfee's maxFee)");
+        }
+        if (options.exists("force")) {
+            fForce = options["force"].get_bool();
+        }
+    }
+
+    // calculate the old fee and fee-rate
+    CAmount nOldFee = wtx.GetDebit(ISMINE_SPENDABLE) - wtx.GetValueOut();
+    int64_t txSize = GetVirtualTransactionSize((CTransaction)wtx);
+    CFeeRate nOldFeeRate(nOldFee, txSize);
+
+    // use the user-defined payTxFee if possible, otherwise use smartfee / fallbackfee
+    CAmount nNewFee = 0;
+    CFeeRate nNewFeeRate = payTxFee;
+    if (nNewFeeRate.GetFeePerK() == 0)
+        nNewFeeRate = mempool.estimateSmartFee(newConfirmTarget);
+    if (nNewFeeRate.GetFeePerK() == 0)
+        nNewFeeRate = CWallet::fallbackFee;
+
+    // new fee rate must be at least old rate + minimum relay rate
+    if (nNewFeeRate.GetFeePerK() < nOldFeeRate.GetFeePerK() + ::minRelayTxFee.GetFeePerK()) {
+        nNewFeeRate = CFeeRate(nOldFeeRate.GetFeePerK() + ::minRelayTxFee.GetFeePerK());
+    }
+
+    // signature sizes can vary by a byte, so add 1 for each input when calculating the new fee
+    nNewFee = nNewFeeRate.GetFee(txSize + wtx.vin.size());
+
+    // if user set totalFee, use that instead
+    if (totalFee > 0) {
+        CAmount minTotalFee = nOldFeeRate.GetFee(txSize+wtx.vin.size()) + minRelayTxFee.GetFee(txSize+wtx.vin.size());
```

Comment:
You should calculate txSize + wtx.vin.size() once and keep it in a const variable for these...

```C++
const int64_t maxNewTxSize = txSize + wtx.vin.size();
```